// @hubspot/ui-extensions/dist/__synced__/crmRemoteComponents.synced.js
// Do not manually update this file, changes will be autogenerated by our scripts based on: ui-extensions-remote-renderer/static/js/crmRemoteComponents.ts
import { createRemoteReactComponent } from '@remote-ui/react';
export const CrmPropertyList = createRemoteReactComponent('CrmPropertyList');
export const CrmAssociationTable = createRemoteReactComponent('CrmAssociationTable');
export const CrmDataHighlight = createRemoteReactComponent('CrmDataHighlight');
export const CrmReport = createRemoteReactComponent('CrmReport');
export const CrmAssociationPivot = createRemoteReactComponent('CrmAssociationPivot');
export const CrmAssociationPropertyList = createRemoteReactComponent('CrmAssociationPropertyList');
export const CrmAssociationStageTracker = createRemoteReactComponent('CrmAssociationStageTracker');
export const CrmSimpleDeadline = createRemoteReactComponent('CrmSimpleDeadline');
export const CrmStageTracker = createRemoteReactComponent('CrmStageTracker');
export const CrmStatistics = createRemoteReactComponent('CrmStatistics');
export const CrmActionButton = createRemoteReactComponent('CrmActionButton');
export const CrmActionLink = createRemoteReactComponent('CrmActionLink');
export const CrmCardActions = createRemoteReactComponent('CrmCardActions');
