// @hubspot/ui-extensions/dist/__synced__/experimentalRemoteComponents.synced.js
// Do not manually update this file, changes will be autogenerated by our scripts based on: ui-extensions-remote-renderer/static/js/experimentalRemoteComponents.ts
import { createRemoteReactComponent } from '@remote-ui/react';
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Iframe = createRemoteReactComponent('Iframe');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const MediaObject = createRemoteReactComponent('MediaObject', {
    fragmentProps: ['itemRight', 'itemLeft'],
});
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Stack2 = createRemoteReactComponent('Stack2');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Center = createRemoteReactComponent('Center');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Grid = createRemoteReactComponent('Grid');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const GridItem = createRemoteReactComponent('GridItem');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const SettingsView = createRemoteReactComponent('SettingsView');
/**
 * The `ExpandableText` component renders a text that can be expanded or collapsed based on a maximum height.
 *
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/expandable-text ExpandableText Docs}
 */
export const ExpandableText = createRemoteReactComponent('ExpandableText');
/**
 * The `Popover` component renders a popover overlay that can contain other components.
 *
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/popover Popover Docs}
 */
export const Popover = createRemoteReactComponent('Popover');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const FileInput = createRemoteReactComponent('FileInput');
