// @hubspot/ui-extensions/dist/experimental/testing/internal/fragment.js
import { RenderedNodeType } from '../types';
import { EMPTY_CHILDREN } from './constants';
import { debugLog } from './debug';
import { printNode } from './print';
import { find, findAll, findAllChildren, findChild, maybeFind, maybeFindChild, } from './query';
/**
 * Creates a rendered fragment node.
 *
 * @param document The document that the fragment belongs to.
 * @returns A rendered fragment node.
 */
export const createFragmentNode = (document) => {
    const fragmentNode = {
        nodeType: RenderedNodeType.Fragment,
        text: null,
        children: EMPTY_CHILDREN,
        document,
        debugLog: (label) => {
            debugLog(fragmentNode, label);
        },
        find: (targetComponent, matcher) => {
            return find(fragmentNode, targetComponent, matcher);
        },
        findAll: (targetComponent, matcher) => {
            return findAll(fragmentNode, targetComponent, matcher);
        },
        findChild: (targetComponent, matcher) => {
            return findChild(fragmentNode, targetComponent, matcher);
        },
        findAllChildren: (targetComponent, matcher) => {
            return findAllChildren(fragmentNode, targetComponent, matcher);
        },
        maybeFind: (targetComponent, matcher) => {
            return maybeFind(fragmentNode, targetComponent, matcher);
        },
        maybeFindChild: (targetComponent, matcher) => {
            return maybeFindChild(fragmentNode, targetComponent, matcher);
        },
        toString: () => {
            return printNode(fragmentNode);
        },
    };
    return fragmentNode;
};
